proposals/NNNN-method-keypaths.md

```diff
@@ -0,0 +1,667 @@
+end of a longer chain involving properties, subscripts, and optional chaining / unwrapping:
+
+```swift
+URL.baseURL?.pathComponents[1].reversed   // compile error: no chains allowed here, only one member deep
+\URL.baseURL?.pathComponents[1].reversed  // OK!
+```
+
+Note that because (1) methods in key paths return functions and (2) functions themselves have neither members nor
+subscripts, methods can only appear as the last element in a key path. This could change if a future proposal for
+example allowed embedding of method arguments in a key path, or somehow allowed function types to conform to protocols.
+
+---
+
+Since unbound methods are never properties, and key paths are currently never methods, this proposal introduces a
+conflict when a method and a property have the same name, e.g. `Array.first` and `Array.first(where:)`. In this case, a
+bare reference with no argument names always refers to the property, and argument names are necessary to refer to the
+method:
+
+```swift
+\[Int].first          // keypath to property
+\[Int].first(where:)  // keypath to method
+```
+
+This parallels the existing behavior of bound but unapplied methods, which face the same problem and solve it in the
+same way:
+
+```swift
+var a: [Int]
+
+a.first          // property
+a.first(where:)  // unapplied method
+```
+
+This rule guarantees that it is always possible to reference both the property and the methods via key paths, since it
+is a compile-time error to declare a no-args method with the same name as a property:
+
+```swift
+struct S {
+    var x: Int
+    func x(_: Int) { … }   // allowed
+    func x() { … }         // compiler error
+}
+```
+
+
+## Source compatibility
+
+This is a pure additive proposal: it only concerns key paths that end with method references, which currently do not
+compile at all. In the case where a funciton and a property have the same same, any existing key path that compiles
+today will continue to refer to the property.
+
+
+## Effect on ABI stability
+
+**TODO**: _I don't think this affects the ABI, since the generated types (`KeyPath<Base,(T)->U>`) are representible in
+Swift’s current type system even if the key path syntax doesn't support them. Are there ABI implications beyond that?
+I’m not even sure what ABI impacts look like. -PPC_
+
+
+## Effect on API resilience
+
+APIs might choose to expose or require key path types with function values, e.g. `KeyPath<String, () -> Int>`, as a
+result of this proposal. However, these types are already legal in Swift; this proposal simply provides a new way to
+construct such values.
+
+
+## Future directions
+
+Some popular feature ideas are out of this proposal’s scope. The discussion below explains why.
+
+### Method calls in key paths
+
+This proposal deals only with _unapplied_ methods. There is also a strong case to make for supporting _applied_
+methods, i.e. allowing key paths to embed method calls.
```

Comment:
i.e. takes a comma

Reply:
Where? Again, I don't see where one is missing.

Reply:
Ah, I see that this is a matter of varying opinion: https://www.dailywritingtips.com/comma-after-i-e-and-e-g/

I've always elided the comma after. TIL.

Reply:
It says "i.e. allowing..." and should say "i.e., allowing..."

I haven't quite gotten to them but your e.g.'s should also take a comma.